### PR TITLE
Add support for authentication based on reverse proxy email

### LINKED ISSUE #19948: Add support for reverse proxy header authentication with email alone
### Feature Description

Hi!

I saw it's possible to authenticate using a reverse proxy by sending an `X-Webauth-User` header. Optionally (if enabled in settings), if the `X-Webauth-Email` header is provided a user is also auto-registered.

For my internal setup at home, my reverse proxy does not have access to gitea usernames, but the sign in system has access to emails. Would it be possible to extend the code to allow authenticating just via the contents of `X-Webauth-Email`?

I'm happy to try and put up a PR myself if there are no objections to this feature.

### Screenshots

_No response_

### PR DESCRIPTION
**Pull request by mhlakhani**
This is useful in scenarios where the reverse proxy may have knowledge
of user emails, but does not know about usernames set on gitea,
as in the feature request in #19948.

I tested this by setting up a fresh gitea install with one user `mhl`
and email `m.hasnain.lakhani@gmail.com`. I then created a private repo,
and configured gitea to allow reverse proxy authentication.

Via curl I confirmed that these two requests now work and return 200s:

curl http://localhost:3000/mhl/private -I --header "X-Webauth-User: mhl"
curl http://localhost:3000/mhl/private -I --header "X-Webauth-Email: m.hasnain.lakhani@gmail.com"

Before this commit, the second request did not work.

I also verified that if I provide an invalid email or user,
a 404 is correctly returned as before

Closes #19948

### CODE REVIEW

File: services/auth/reverseproxy.go [hunk 1]
```diff
@@ -71,6 +71,58 @@ func (r *ReverseProxy) Verify(req *http.Request, w http.ResponseWriter, store Da
...
+	return user
+}
+
+// getEmail extracts the email from the "setting.ReverseProxyAuthEmail" header
+func (r *ReverseProxy) getEmail(req *http.Request) string {
+	webAuthEmail := strings.TrimSpace(req.Header.Get(setting.ReverseProxyAuthEmail))
+	if len(webAuthEmail) == 0 {
+		return ""
+	}
+	return webAuthEmail
```
Comment (Gusted) [inline services/auth/reverseproxy.go]: Same effect.
```suggestion
	return strings.TrimSpace(req.Header.Get(setting.ReverseProxyAuthEmail))
```

Comment (mhlakhani) [reply services/auth/reverseproxy.go]: This was copied from https://github.com/go-gitea/gitea/blob/3ba09103a4bd35f6e24eecbc85658bdbf404092f/services/auth/reverseproxy.go#L40-L44 - should I also change that?

Comment (Gusted) [reply services/auth/reverseproxy.go]: Yes.

**Review by Gusted — APPROVED** [review]

**Comment by mhlakhani** [comment]
hi, is there anything I need to do to get the other +1 so this PR can be pushed? Hoping to use a "real" version of gitea rather than one I've self built :)

**Review by lunny — APPROVED** [review]

**Review by 6543 — APPROVED** [review]

**Pull request merged**